### PR TITLE
UniqueTagList: Add assertions to check all tags are unique #411

### LINKED ISSUE #411: Add assertions to check UniqueTagList’s invariant
The invariant of `UniqueTagList` (i.e that all the tags in `UniqueTagList` are unique) is not being checked before and after the call of every public method in `UniqueTagList`.

From: https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html, assertions should be used to check for internal invariants.

Let's create assertions to check for UniqueTagList's invariant.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #411

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/1] [Add assertions to check UniqueTagList’s invariant](https://github.com/se-edu/addressbook-level4/pull/412/commits/99bd6559c7f4f9f7b169632889cbc23d58a5c981)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/0f2cb22b3dbf7cdc592e2b58261728241f4d4166...99bd6559c7f4f9f7b169632889cbc23d58a5c981))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/412/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/1] [UniqueTagList: Add assertions to check all tags are unique](https://github.com/se-edu/addressbook-level4/pull/412/commits/8d94d0c5706aaf7be00901c544f941057e118e14)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/0f2cb22b3dbf7cdc592e2b58261728241f4d4166...8d94d0c5706aaf7be00901c544f941057e118e14)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/a62f298c8828d3fa8e37743c58a3f1b92ba91273~1...a62f298c8828d3fa8e37743c58a3f1b92ba91273))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/412/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/1] [UniqueTagList: Add assertions to check all tags are unique](https://github.com/se-edu/addressbook-level4/pull/412/commits/cb215966208f2a997c36b500c69340bc104da3be)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/0f2cb22b3dbf7cdc592e2b58261728241f4d4166...cb215966208f2a997c36b500c69340bc104da3be)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/b5d1f27168c81522b4f8bfeb8693bbbcd0917b28~1...b5d1f27168c81522b4f8bfeb8693bbbcd0917b28))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/412/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 1]
```diff
@@ -81,13 +89,16 @@ public UniqueTagList(Set<Tag> tags) {
...
+
+        assert CollectionUtil.elementsAreUnique(internalList);
     }
 
     /**
      * Returns all tags in this list as a Set.
      * This set is mutable and change-insulated against the internal list.
      */
     public Set<Tag> toSet() {
+        assert CollectionUtil.elementsAreUnique(internalList);
```
Comment (eugenepeh) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: is the assert necessary here? This method is similar to getters, if the checks has been done at all setters, is it still necessary to perform the check at getters?

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 2]
```diff
@@ -135,30 +151,38 @@ public void add(Tag toAdd) throws DuplicateTagException {
...
 
     public boolean equalsOrderInsensitive(UniqueTagList other) {
+        assert CollectionUtil.elementsAreUnique(internalList);
+        assert CollectionUtil.elementsAreUnique(other.internalList);
         return this == other || new HashSet<>(this.internalList).equals(new HashSet<>(other.internalList));
     }
 
     @Override
     public int hashCode() {
+        assert CollectionUtil.elementsAreUnique(internalList);
```
Comment (MightyCupcakes) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: isnt this too excessive? Why would a check need to be done here if the internal list did not even change

**Comment by Zhiyuan-Amos** [comment]
@eugenepeh @MightyCupcakes will reply both of you here since your comments are similar:

Yes, it is true that if methods that edit the state of `UniqueTagList` already has assertions to ensure that the uniqueness of each tag in `UniqueTagList` is maintained, then it makes logical sense that the getter methods don't need to be checked. 

Sounds lame, but there's the possibility of save data corruption / future programmers adding in new method but forgetting to perform the invariant check etc.
Then having assertions for getter methods is helpful to prevent further execution of the code (since the program will crash). The callers of the getter methods **are expecting a valid UniqueTagList**, thus if the **UniqueTagList isn't valid**, the operations further down may be affected. Assertions over here basically stop further execution of code to prevent further issues from happening.

**Comment by yamgent** [comment]
Languages that support class invariant (such as Eiffel) actually enforces their invariant checks for every public method that the class has.

Perhaps an extremely good compiler that supports a 'design-by-contract' language would possibly eliminate unnecessary invariant checks, if it can guarantee that the variables are not modified when a call is made.

Unfortunately, class invariant and 'design-by-contract' is not a native feature of Java. So hopefully our implementation is a good enough emulation of invariant (additional comments about this portion of the PR are welcome).

**Comment by damithc** [comment]
Also keep in mind that we are using `asserts` that aren't going to be enabled by typical users (there is no way to enable assertions for jar files, unless the user runs it in command line). So, these statements are executed only when devs test the app.

**Pull request merged**